placement/algo/placement.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package algo
+
+import (
+	"encoding/json"
+	"errors"
+	"sort"
+
+	"github.com/m3db/m3cluster/placement"
+)
+
+var (
+	errShardsWithDifferentReplicas = errors.New("invalid placement, found shards with different replicas")
+)
+
+// placementSnapshot implements placement.Snapshot
+type placementSnapshot struct {
```

Comment:
yeah so I feel like defining an interface even for internal structs narrows the access surface and hides implementation details so you can swap things out underneath quite easily. Right now we are accessing the internal fields of the structs fairly liberally so it's fairly difficult to make sense of the coupling between `placementSnapshot` and its users.